src/lib/Microsoft.Health.Fhir.Ingest.Template/CalculatedFunction/MeasurementExtractor.cs

```diff
@@ -0,0 +1,166 @@
+            }
+
+            return default;
+        }
+
+        protected static bool IsExpressionDefined(params Expression[] expressions)
+        {
+            EnsureArg.IsNotNull(expressions, nameof(expressions));
+
+            return expressions.Any(ex => ex != null && !string.IsNullOrWhiteSpace(ex.Value));
+        }
+
+        protected virtual DateTime? GetTimestamp(JToken token) => EvalExpression<DateTime?>(token, nameof(Template.TimestampExpression), true, Template.TimestampExpression);
+
+        protected virtual string GetDeviceId(JToken token) => EvalExpression<string>(token, nameof(Template.DeviceIdExpression), true, Template.DeviceIdExpression);
+
+        protected virtual string GetPatientId(JToken token) => EvalExpression<string>(token, nameof(Template.PatientIdExpression), false, Template.PatientIdExpression);
+
+        protected virtual string GetEncounterId(JToken token) => EvalExpression<string>(token, nameof(Template.EncounterIdExpression), false, Template.EncounterIdExpression);
+
+        protected virtual string GetCorrelationId(JToken token) => EvalExpression<string>(token, nameof(Template.CorrelationIdExpression), true, Template.CorrelationIdExpression);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        protected virtual IEnumerable<JToken> MatchTypeTokens(JObject token)
+        {
+            EnsureArg.IsNotNull(token, nameof(token));
+            var evaluator = ExpressionEvaluatorFactory.Create(Template.TypeMatchExpression);
+
+            foreach (var extractedToken in evaluator.SelectTokens(token))
+            {
+                // Add the extracted data as an element of the original data.
+                // This allows subsequent expressions access to data from the original event data
+
+                var tokenClone = token.DeepClone() as JObject;
+                tokenClone.Add(MatchedToken, extractedToken);
+                yield return tokenClone;
+            }
+        }
+
+        private Measurement CreateMeasurementFromToken(JToken token)
+        {
+            // Current assumption is that the expressions should match a single element and will error otherwise.
+
+            string deviceId = GetDeviceId(token);
+            DateTime? timestamp = GetTimestamp(token);
+
+            string correlationId = null;
+            if (IsExpressionDefined(Template.CorrelationIdExpression))
+            {
+                correlationId = GetCorrelationId(token);
+            }
+
+            var measurement = new Measurement
+            {
+                DeviceId = deviceId,
+                OccurrenceTimeUtc = timestamp.Value,
+                Type = Template.TypeName,
+                PatientId = GetPatientId(token),
+                EncounterId = GetEncounterId(token),
+                CorrelationId = correlationId,
+            };
+
+            if (Template.Values != null)
+            {
+                for (var i = 0; i < Template.Values.Count; i++)
+                {
```

Comment:
Is there a reason we can't just use "foreach" here?

Reply:
Previously I was constructing the name of the expression using `ValueExpression{idx}` but  then realized each ValueExpression had its own name. I'll switch to use a `foreach`.